<a href="https://colab.research.google.com/github/AyleenSantander/M.11_Intro_AI/blob/main/M_11_A_11_Data_Processing_Python_Dvlt_and_Pds.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Assignment 4: Data Processing - Python Development and Pandas
#Name Ayleen Santander

#Part 1: Create a DataFrame:
1. Create a DataFrame with the given data for Name, Age, and City.
2. Ensure the column names are correct and the data types are appropriate.

In [ ]:
#Libraries
import sys
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
from functools import reduce
%matplotlib inline

In [ ]:
df = pd.DataFrame({
    'Name': ['Mateo', 'Priya', 'Jonas', 'Amara', 'Liam'],
    'Age': [31, 27, 45, 38, 24],
    'City': ['Toronto', 'Mumbai', 'Berlin', 'Lagos', 'Dublin']
})
df

,Name,Age,City
0,Mateo,31,Toronto
1,Priya,27,Mumbai
2,Jonas,45,Berlin
3,Amara,38,Lagos
4,Liam,24,Dublin


In [ ]:
print(df.info())
print(df.shape)
print(df.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    5 non-null      object
 1   Age     5 non-null      int64 
 2   City    5 non-null      object
dtypes: int64(1), object(2)
memory usage: 252.0+ bytes
None
(5, 3)
Name    object
Age      int64
City    object
dtype: object


#Part 2. Row and Column Manipulation:
1. Successfully drop the 'City' column from the DataFrame.
2. Verify that the resulting DataFrame only has 'Name' and 'Age' columns.

In [ ]:
df= df.drop(columns=['City'])
df

,Name,Age
0,Mateo,31
1,Priya,27
2,Jonas,45
3,Amara,38
4,Liam,24


#Part 3. Handling Null Values:
1. Create a DataFrame with at least one null value in each column.
2. Fill all null values with an appropriate value (e.g., 0 for numeric columns).
3. Demonstrate understanding of different methods to handle null values.

In [ ]:
#1. Create a DataFrame with at least one null value in each column.
df_null = pd.DataFrame({
    'Employee': ['Sofia', 'Karim', 'Elena', 'Noah', 'Rina'],
    'Years': [6.0, np.nan, 12.0, 3.0, np.nan],
    'Country': ['Spain', 'Morocco', np.nan, 'Canada', 'Japan'],
    'Bonus': [4500, 3200, np.nan, 2800, 5100]
})
print(df_null)
print('\nNulls per column:\n', df_null.isnull().sum())
print('\nTotal nulls:', df_null.isnull().sum().sum())

  Employee  Years  Country   Bonus
0    Sofia    6.0    Spain  4500.0
1    Karim    NaN  Morocco  3200.0
2    Elena   12.0      NaN     NaN
3     Noah    3.0   Canada  2800.0
4     Rina    NaN    Japan  5100.0

Nulls per column:
 Employee    0
Years       2
Country     1
Bonus       1
dtype: int64

Total nulls: 4


In [ ]:
#2. Fill all null values with an appropriate value (e.g., 0 for numeric columns).
df_null= df_null.fillna(1)
# df_null = df_null.fillna(df_null.mean(numeric_only=True))
print(df_null)
print('\nNulls per column:\n', df_null.isnull().sum())
print('\nTotal nulls:', df_null.isnull().sum().sum())

  Employee  Years  Country   Bonus
0    Sofia    6.0    Spain  4500.0
1    Karim    1.0  Morocco  3200.0
2    Elena   12.0        1     1.0
3     Noah    3.0   Canada  2800.0
4     Rina    1.0    Japan  5100.0

Nulls per column:
 Employee    0
Years       0
Country     0
Bonus       0
dtype: int64

Total nulls: 0


#Part 4: GroupBy and Describe:
1. Use GroupBy to group the data by 'Category'.
2. Apply the describe() function to the 'Value' column.
3. Interpret the resulting statistics for each category.

In [ ]:
data2 = {
    'Category': ['Electronics', 'Clothing', 'Electronics', 'Furniture',
                 'Clothing', 'Electronics', 'Furniture', 'Clothing'],
    'Product':  ['Laptop', 'T-Shirt', 'Phone', 'Chair',
                 'Jeans', 'Tablet', 'Desk', 'Jacket'],
    'Sales':    [1200, 25, 800, 150, 60, 450, 300, 90],
    'Quantity': [3, 20, 5, 8, 12, 6, 4, 7]
}

df = pd.DataFrame(data2)
print(df)

      Category  Product  Sales  Quantity
0  Electronics   Laptop   1200         3
1     Clothing  T-Shirt     25        20
2  Electronics    Phone    800         5
3    Furniture    Chair    150         8
4     Clothing    Jeans     60        12
5  Electronics   Tablet    450         6
6    Furniture     Desk    300         4
7     Clothing   Jacket     90         7


In [ ]:
grouped = df.groupby('Category')

# single aggregation
print(grouped['Sales'].sum())

# multiple stats on one column
print(grouped['Sales'].agg(['sum', 'mean', 'count']))

# different stats per column
print(grouped.agg({'Sales': 'sum', 'Quantity': 'mean'}))


Category
Clothing        175
Electronics    2450
Furniture       450
Name: Sales, dtype: int64
              sum        mean  count
Category                            
Clothing      175   58.333333      3
Electronics  2450  816.666667      3
Furniture     450  225.000000      2
             Sales   Quantity
Category                     
Clothing       175  13.000000
Electronics   2450   4.666667
Furniture      450   6.000000


In [ ]:
#2. Apply the describe() function to the 'Value' column.
print(grouped[['Sales', 'Quantity']].describe())

            Sales                                                       \
            count        mean         std    min    25%    50%     75%   
Category                                                                 
Clothing      3.0   58.333333   32.532035   25.0   42.5   60.0    75.0   
Electronics   3.0  816.666667  375.277675  450.0  625.0  800.0  1000.0   
Furniture     2.0  225.000000  106.066017  150.0  187.5  225.0   262.5   

                    Quantity                                                   
                max    count       mean       std  min  25%   50%   75%   max  
Category                                                                       
Clothing       90.0      3.0  13.000000  6.557439  7.0  9.5  12.0  16.0  20.0  
Electronics  1200.0      3.0   4.666667  1.527525  3.0  4.0   5.0   5.5   6.0  
Furniture     300.0      2.0   6.000000  2.828427  4.0  5.0   6.0   7.0   8.0  


#Interpretation of resulting statistcs
Sales results
Electronics = 3 products, averaging 816.67 per sale. Range from 450 to at least 1000 at the 75th percentile. Highest-value category by a wide margin, roughly 14× Clothing's average.

Clothing = 3 products, averaging 58.33. Median of 60 sits close to the mean, so the values are fairly evenly spread rather than dragged by one outlier.

Furniture = only 2 products, averaging 225. With n=2, the median (225) is just the midpoint of the two values, and the 25%/75% marks are interpolated between them rather than describing any real spread.

Quantity/Stock results
Clothing = mean 13 units, range 7–20. Highest volume.
Furniture = mean 6, range 4–8.
Electronics = mean 4.67, range 3–6. Lowest volume, and the tightest spread (std 1.53).

#Part 5: Concatenation and Merging:
1. Concatenate df1 and df2 vertically.
2. Merge the result with df3 horizontally.
3 .Ensure the resulting DataFrame has the correct shape and all data.

In [ ]:
#1. Concatenate df1 and df2 vertically
df1 = pd.DataFrame({
    'Category': ['Electronics', 'Clothing', 'Furniture'],
    'Product':  ['Laptop', 'T-Shirt', 'Chair'],
    'Sales':    [1200, 25, 150]
})

df2 = pd.DataFrame({
    'Category': ['Electronics', 'Clothing', 'Furniture'],
    'Product':  ['Phone', 'Jeans', 'Desk'],
    'Sales':    [800, 60, 300]
})
df_combined = pd.concat([df1, df2], ignore_index=True)
print(df_combined)

      Category  Product  Sales
0  Electronics   Laptop   1200
1     Clothing  T-Shirt     25
2    Furniture    Chair    150
3  Electronics    Phone    800
4     Clothing    Jeans     60
5    Furniture     Desk    300


In [ ]:
df3 = pd.DataFrame({
    'Category': ['Electronics', 'Clothing', 'Furniture'],
    'Manager':  ['Ana', 'Luis', 'Marta'],
    'Region':   ['North', 'South', 'East']
})

In [ ]:
#2. Merge the result with df3 horizontally
df_merged = pd.merge(df_combined, df3, on='Category', how='left')
print(df_merged)

      Category  Product  Sales Manager Region
0  Electronics   Laptop   1200     Ana  North
1     Clothing  T-Shirt     25    Luis  South
2    Furniture    Chair    150   Marta   East
3  Electronics    Phone    800     Ana  North
4     Clothing    Jeans     60    Luis  South
5    Furniture     Desk    300   Marta   East


#Part 6: Tuples and Sets:
1. Create a tuple of at least 3 fruits and a set of at least 5 numbers.
2. Attempt to add an element to both the tuple and the set.
3. Explain the difference in behavior between tuples and sets.

In [ ]:
#1. Create a tuple of at least 3 fruits and a set of at least 5 numbers.
fruits = ('orange', 'banana', 'strawberry', 'blackberry', 'watermelon')
numbers = {3, 7, 12, 18, 25}

print(fruits, type(fruits))
print(numbers, type(numbers))

('orange', 'banana', 'strawberry', 'blackberry', 'watermelon') <class 'tuple'>
{18, 3, 7, 25, 12} <class 'set'>


In [ ]:
#2. Attempt to add an element to both the tuple and the set.
# Set — works
numbers.add(30)
print("Set after add:", numbers)

# Tuple — fails
try:
    fruits.append('pear')
except AttributeError as e:
    print("Tuple error:", e)

Set after add: {18, 3, 7, 25, 12, 30}
Tuple error: 'tuple' object has no attribute 'append'


Explanation:
- Tuples are inmutable so after create a tuples we cannot add or change
- Sets are unorded collection of unique element we can identify using the set{}

#Part 7: Dictionaries:
1. Create a dictionary with at least 3 student names and their scores.
2. Update the score of one existing student.
3. Add a new student with their score to the dictionary.

In [ ]:
#1. Create a dictionary with at least 3 student names and their scores.
students = {
    'Ayleen': 92,
    'Diego': 85,
    'Marta': 78
}

print(students)
print(type(students))

{'Ayleen': 92, 'Diego': 85, 'Marta': 78}
<class 'dict'>


In [ ]:
print("Before:", students['Diego'])


Before: 85


In [ ]:
#2. Update the score of one existing student.
students['Diego'] = 90
print(students)

{'Ayleen': 92, 'Diego': 90, 'Marta': 78}


In [ ]:
print("After:", students['Diego'])

After: 90


In [ ]:
#3. Add a new student with their score to the dictionary.
students['Luis'] = 88
print(students)

In [ ]:
#Another way to apply the same concepts
if 'Luis' not in students:
    students['Luis'] = 88
    print("Added")
else:
    print("Already exists — would overwrite")

Added


#Part 8:Functions and Lambda:
1. Define a regular function to calculate the square of a number.
2. Create a lambda function to perform the same operation.
3. Demonstrate the use of both functions with at least two different inputs.

In [ ]:
#1. Define a regular function to calculate the square of a number.
def square(number):
    return number ** 2

print(square(5))
print(square(2.5))
print(square(-3))

25
6.25
9


In [ ]:
#2.Create a lambda function to perform the same operation.
square_lambda = lambda number: number ** 2

print(square_lambda(10))
print(square_lambda(22.5))

100
506.25


In [ ]:
#other example
numbers = [1, 2, 3, 4, 5]

squared = list(map(lambda x: x ** 2, numbers))
print(squared)

[1, 4, 9, 16, 25]


In [ ]:
#3. Demonstrate the use of both functions with at least two different inputs.
def square(number):
    return number ** 2

square_lambda = lambda number : number ** 2

# Input 1
print("Input: 4")
print("  Regular function:", square(4))
print("  Lambda function: ", square_lambda(4))

# Input 2
print("Input: 7")
print("  Regular function:", square(7))
print("  Lambda function: ", square_lambda(7))

Input: 4
  Regular function: 16
  Lambda function:  16
Input: 7
  Regular function: 49
  Lambda function:  49


#Part 9: Iterators and Generators:
1. Create an iterator class that generates the first 5 even numbers.
2. Create a generator function that does the same.
3. Demonstrate the use of both the iterator and generator.

In [ ]:
#1. Create an iterator class that generates the first 5 even numbers.
class EvenNumbers:
    def __init__(self, count=5):
        self.count = count
        self.current = 0
        self.generated = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.generated >= self.count:
            raise StopIteration
        self.current += 2
        self.generated += 1
        return self.current


evens = EvenNumbers(5)

for number in evens:
    print(number)

2
4
6
8
10


In [ ]:
#2. Create a generator function that does the same.
def even_numbers(count=5):
    current = 0
    for _ in range(count):
        current += 2
        yield current


for number in even_numbers(5):
    print(number)

2
4
6
8
10


In [ ]:
#3. Demonstrate the use of both the iterator and generator.
class EvenNumbers:
    def __init__(self, count=5):
        self.count = count
        self.current = 0
        self.generated = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.generated >= self.count:
            raise StopIteration
        self.current += 2
        self.generated += 1
        return self.current


def even_numbers(count=5):
    current = 0
    for _ in range(count):
        current += 2
        yield current


# --- Iterator class ---
print("Iterator class:")
for number in EvenNumbers(5):
    print(" ", number)

# --- Generator function ---
print("\nGenerator function:")
for number in even_numbers(5):
    print(" ", number)

# --- Same output? ---
print("\nAs lists:")
print("  Iterator: ", list(EvenNumbers(5)))
print("  Generator:", list(even_numbers(5)))
print("  Match:", list(EvenNumbers(5)) == list(even_numbers(5)))

Iterator class:
  2
  4
  6
  8
  10

Generator function:
  2
  4
  6
  8
  10

As lists:
  Iterator:  [2, 4, 6, 8, 10]
  Generator: [2, 4, 6, 8, 10]
  Match: True


#Part 10: Map, Reduce, and Filter:
1. Use map() to square all numbers in the given list.
2. Use reduce() to find the product of all numbers in the list.
3. Use filter() to get only the even numbers from the list.
4. Print the results of each operation.

In [ ]:
#1. Use map() to square all numbers in the given list.
numbers = [1, 2, 3, 4, 5]

squared = list(map(lambda x: x ** 2, numbers))
print(squared)

[1, 4, 9, 16, 25]


In [ ]:
#2. Use reduce() to find the product of all numbers in the list.
numbers = [1, 2, 3, 4, 5]

product = reduce(lambda x, y: x * y, numbers)
print(product)   # 120

120


In [ ]:
#3. Use filter() to get only the even numbers from the list.
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print(even_numbers)

[2, 4, 6, 8, 10]


#Part 11: Object-Oriented Programming:
1. Create a Rectangle class with attributes for length and width.
2. Implement methods to calculate area and perimeter.
3. Create at least two Rectangle objects and demonstrate the use of their methods.

#github link M_11_A_11_Data_Processing_Python_Dvlt_and_Pds.ipynb
#colab notebook link https://colab.research.google.com/drive/1ls2xR30wLHM3HgM0Ulipo1jFlBSpVPeS?usp=sharing

In [ ]:
#1. Create a Rectangle class with attributes for length and width.
class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width


rect = Rectangle(10, 5)

print(rect.length)
print(rect.width)

10
5


In [ ]:
#2. Implement methods to calculate area and perimeter.
class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):
        return self.length * self.width

    def perimeter(self):
        return 2 * (self.length + self.width)


rect = Rectangle(10, 5)

print("Area:", rect.area())
print("Perimeter:", rect.perimeter())

Area: 50
Perimeter: 30


In [ ]:
#3. Create at least two Rectangle objects and demonstrate the use of their methods.
class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):
        return self.length * self.width

    def perimeter(self):
        return 2 * (self.length + self.width)


rect1 = Rectangle(10, 5)
rect2 = Rectangle(3, 7)

print("Rectangle 1")
print("  Length:", rect1.length, " Width:", rect1.width)
print("  Area:", rect1.area())
print("  Perimeter:", rect1.perimeter())

print("Rectangle 2")
print("  Length:", rect2.length, " Width:", rect2.width)
print("  Area:", rect2.area())
print("  Perimeter:", rect2.perimeter())

Rectangle 1
  Length: 10  Width: 5
  Area: 50
  Perimeter: 30
Rectangle 2
  Length: 3  Width: 7
  Area: 21
  Perimeter: 20


#Part 12: Pandas Data Analysis:
1. Calculate and display the average salary by department.
2. Extract and display the names of employees with salary > 60000.
3. Add a 'Bonus' column that is 10% of the salary.
4. Ensure all operations are performed on the DataFrame correctly.

In [ ]:
#1. Calculate and display the average salary by department.

HR = {
    'Employee':   ['Ana', 'Luis', 'Marta', 'Diego', 'Sofia', 'Carlos'],
    'Department': ['Sales', 'IT', 'Sales', 'HR', 'IT', 'HR'],
    'Salary':     [55000, 72000, 61000, 48000, 80000, 52000]
}

df = pd.DataFrame(HR)

avg_salary = df.groupby('Department')['Salary'].mean()
print(avg_salary)

Department
HR       50000.0
IT       76000.0
Sales    58000.0
Name: Salary, dtype: float64


In [ ]:
#2. Extract and display the names of employees with salary > 60000.
high_salary = df[df['Salary'] > 60000]['Employee']
print(high_salary)

1     Luis
2    Marta
4    Sofia
Name: Employee, dtype: object


In [ ]:
#3. Add a 'Bonus' column that is 10% of the salary.
df['Bonus'] = df['Salary'] * 0.10
print(df)

  Employee Department  Salary   Bonus
0      Ana      Sales   55000  5500.0
1     Luis         IT   72000  7200.0
2    Marta      Sales   61000  6100.0
3    Diego         HR   48000  4800.0
4    Sofia         IT   80000  8000.0
5   Carlos         HR   52000  5200.0


In [ ]:
#4. Ensure all operations are performed on the DataFrame correctly.
print("Shape:", df.shape)
print("Columns:", df.columns.tolist())
print()

print("Data types:")
print(df.dtypes)
print()

print("Missing values:")
print(df.isnull().sum())
print()

print("Full DataFrame:")
print(df)

Shape: (6, 4)
Columns: ['Employee', 'Department', 'Salary', 'Bonus']

Data types:
Employee       object
Department     object
Salary          int64
Bonus         float64
dtype: object

Missing values:
Employee      0
Department    0
Salary        0
Bonus         0
dtype: int64

Full DataFrame:
  Employee Department  Salary   Bonus
0      Ana      Sales   55000  5500.0
1     Luis         IT   72000  7200.0
2    Marta      Sales   61000  6100.0
3    Diego         HR   48000  4800.0
4    Sofia         IT   80000  8000.0
5   Carlos         HR   52000  5200.0
